src/xenia/kernel/xboxkrnl/xboxkrnl_rtl.cc

```diff
@@ -188,106 +188,97 @@ void RtlFreeUnicodeString(pointer_t<X_UNICODE_STRING> string) {
+    buffer_ptr[buffer_capacity - 1] = 0;  // \0
   }
-  SHIM_SET_RETURN_32(result);
+  return result;
 }
-
-SHIM_CALL RtlMultiByteToUnicodeN_shim(PPCContext* ppc_context,
-                                      KernelState* kernel_state) {
-  uint32_t destination_ptr = SHIM_GET_ARG_32(0);
-  uint32_t destination_len = SHIM_GET_ARG_32(1);
-  uint32_t written_ptr = SHIM_GET_ARG_32(2);
-  uint32_t source_ptr = SHIM_GET_ARG_32(3);
-  uint32_t source_len = SHIM_GET_ARG_32(4);
-
+DECLARE_XBOXKRNL_EXPORT(RtlUnicodeStringToAnsiString, ExportTag::kImplemented);
+
+// https://msdn.microsoft.com/en-us/library/ff553113
+dword_result_t RtlMultiByteToUnicodeN(pointer_t<uint16_t> destination_ptr,
+                                      dword_t destination_len,
+                                      lpdword_t written_ptr,
+                                      pointer_t<uint8_t> source_ptr,
+                                      dword_t source_len) {
   uint32_t copy_len = destination_len >> 1;
   copy_len = copy_len < source_len ? copy_len : source_len;
 
   // TODO(benvanik): maybe use MultiByteToUnicode on Win32? would require
   // swapping.
 
   for (uint32_t i = 0; i < copy_len; i++) {
-    xe::store_and_swap<uint16_t>(
-        SHIM_MEM_ADDR(destination_ptr + i * 2),
-        xe::load<uint8_t>(SHIM_MEM_ADDR(source_ptr + i)));
+    destination_ptr[i] = source_ptr[i];
   }
 
-  if (written_ptr != 0) {
-    SHIM_SET_MEM_32(written_ptr, copy_len << 1);
+  if (written_ptr.guest_address() != 0) {
+    *written_ptr = copy_len << 1;
   }
 
-  SHIM_SET_RETURN_32(0);
+  return 0;
 }
-
-SHIM_CALL RtlUnicodeToMultiByteN_shim(PPCContext* ppc_context,
-                                      KernelState* kernel_state) {
-  uint32_t destination_ptr = SHIM_GET_ARG_32(0);
-  uint32_t destination_len = SHIM_GET_ARG_32(1);
-  uint32_t written_ptr = SHIM_GET_ARG_32(2);
-  uint32_t source_ptr = SHIM_GET_ARG_32(3);
-  uint32_t source_len = SHIM_GET_ARG_32(4);
-
+DECLARE_XBOXKRNL_EXPORT(RtlMultiByteToUnicodeN,
+                        ExportTag::kImplemented | ExportTag::kSketchy);
+
+// https://msdn.microsoft.com/en-us/library/ff553261
+dword_result_t RtlUnicodeToMultiByteN(pointer_t<uint8_t> destination_ptr,
+                                      dword_t destination_len,
+                                      lpdword_t written_ptr,
+                                      pointer_t<uint16_t> source_ptr,
+                                      dword_t source_len) {
   uint32_t copy_len = source_len >> 1;
   copy_len = copy_len < destination_len ? copy_len : destination_len;
 
   // TODO(benvanik): maybe use UnicodeToMultiByte on Win32?
-
-  auto source = reinterpret_cast<uint16_t*>(SHIM_MEM_ADDR(source_ptr));
-  auto destination = reinterpret_cast<uint8_t*>(SHIM_MEM_ADDR(destination_ptr));
   for (uint32_t i = 0; i < copy_len; i++) {
-    uint16_t c = xe::byte_swap(*source++);
-    *destination++ = c < 256 ? (uint8_t)c : '?';
+    uint16_t c = source_ptr[i];
```

Comment:
Ok, I added the type lpword_t to handle this the same way as lpdword_t, etc... now it is actually doing the byte-swapping